### PR TITLE
Report Outstanding Bundles and Bytes

### PR DESCRIPTION
**Pull request by edman124**
These metrics are needed as a part of an internal project for thread scaling.
The outstanding bundles and bytes will provide signals for how we want to autoscale the workers. The outstanding bundles tracks how many bundles of work are currently being processed (currently very similar to number of active threads). The outstanding bytes tracks how many bytes are currently being processed. These two can be compared to the respective limits to see if the pipeline needs to be scaled up or down.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by edman124** [comment]
ptal @twang126
ptal @m-trieu

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java
Comment (m-trieu) [inline runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:2886]: nit: 10000000
name this something so its clear

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java
Comment (m-trieu) [inline runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:2955]: nit: see above
maybe make this a constant something like `private static final MAXIMUM_BYTES_OUTSTANDING = 10000000`

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java
Comment (m-trieu) [inline runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:2975]: whats the point of `count`?

**Review by m-trieu — CHANGES_REQUESTED** [review]

**Review by twang126 — COMMENTED** [review]
General nit: can you add a bit more color in the description about:

1. What is a bundle. Why is it important
2. What is this bytes referring to. Is it backlog, queued, etc?

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java
Comment (twang126) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:1739]: a little unclear here that bundles are also referred to as elements. Is this supposed to be the case?

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:2975]: It is just some arbitrary work (incrementing a counter) for the threads to execute.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:1739]: The elements was the variable defined already in the BoundedQueueExecutor, should I change this naming?

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:2955]: Fixed.

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:2886]: Fixed.

**Comment by edman124** [comment]
Run Java PreCommit

**Comment by edman124** [comment]
ptal: @bvolpato

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:1739]: I will just leave the naming based on how the BoundedQueueExecutor has named everything.

**Comment by edman124** [comment]
assign set of reviewers

**Comment by Abacn** [comment]
Hi, what is the status of this PR? @edman124  @m-trieu @twang126

It was noted in https://github.com/apache/beam/pull/29065#issuecomment-1771781396 that this PR will fix a flaky test of Java PreCommit that seen quite frequently. We may need this before release -- current Java PreCommit is too flaky

**Comment by kennknowles** [comment]
This flakiness is recent, right? Was there a change that caused it to become flaky? (could be fine but we might want to check)

**Comment by kennknowles** [comment]
It is OK to rollback a change that went in that does not have reliable tests, too.

**Comment by Abacn** [comment]
> This flakiness is recent

Yes. I suggest if this was not made into release cut we should revert #28513

**Comment by edman124** [comment]
I think this PR is ready, @m-trieu can you double check and give a LGTM if everything looks good. I think the PR should be good to submit after approval

**Comment by m-trieu** [comment]
LGTM
reminder to rebase and squash commits

**Review by m-trieu — APPROVED** [review]

**Comment by Abacn** [comment]
thanks, as long as there is no merging conflict rebase is not needed, or is there any recent commit to master branch would affect this PR? Otherwise I will just merge this PR

**Comment by edman124** [comment]
I don't think there are any conflicts

**Comment by AnandInguva** [comment]
@Abacn or I will merge it once tests are green.

**Pull request merged**
